# ARTIFICIAL NEURAL NETWORKS

In [44]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## 1. Data Exploration and Preprocessing

**Begin by loading and exploring the "Alphabets_data.csv" dataset. Summarize its key features such as the number of samples, features, and classes.**

**Execute necessary data preprocessing steps including data normalization, managing missing values.**


In [45]:
df = pd.read_csv('dataset/sonardataset.csv')

In [46]:
df.head()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60,Y
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x_1     208 non-null    float64
 1   x_2     208 non-null    float64
 2   x_3     208 non-null    float64
 3   x_4     208 non-null    float64
 4   x_5     208 non-null    float64
 5   x_6     208 non-null    float64
 6   x_7     208 non-null    float64
 7   x_8     208 non-null    float64
 8   x_9     208 non-null    float64
 9   x_10    208 non-null    float64
 10  x_11    208 non-null    float64
 11  x_12    208 non-null    float64
 12  x_13    208 non-null    float64
 13  x_14    208 non-null    float64
 14  x_15    208 non-null    float64
 15  x_16    208 non-null    float64
 16  x_17    208 non-null    float64
 17  x_18    208 non-null    float64
 18  x_19    208 non-null    float64
 19  x_20    208 non-null    float64
 20  x_21    208 non-null    float64
 21  x_22    208 non-null    float64
 22  x_

In [48]:
df.shape

(208, 61)

In [49]:
df.describe()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_51,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


In [50]:
# Check missing values
print("Missing Values: ")
df.isnull().sum()

Missing Values: 


x_1     0
x_2     0
x_3     0
x_4     0
x_5     0
       ..
x_57    0
x_58    0
x_59    0
x_60    0
Y       0
Length: 61, dtype: int64

In [51]:
print('Class Distribution: \n')
df['Y'].value_counts()

Class Distribution: 



Y
M    111
R     97
Name: count, dtype: int64

* The dataset contains no missing values.

* All 60 input features are continuous numerical variables.

* The target variable is categorical with two classes.

* The dataset is reasonably balanced with 111 mines and 97 rocks.

In [52]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

X = df.drop('Y', axis=1)
y = df['Y']

In [53]:
X.shape, y.shape

((208, 60), (208,))

In [54]:
# Encode target variable
le = LabelEncoder()
y = le.fit_transform(y)

In [55]:
y

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [56]:
# Feature Scaling 
sc = StandardScaler()
X_scaled = sc.fit_transform(X)
X_scaled

array([[-0.39955135, -0.04064823, -0.02692565, ...,  0.06987027,
         0.17167808, -0.65894689],
       [ 0.70353822,  0.42163039,  1.05561832, ..., -0.47240644,
        -0.44455424, -0.41985233],
       [-0.12922901,  0.60106749,  1.72340448, ...,  1.30935987,
         0.25276128,  0.25758223],
       ...,
       [ 1.00438083,  0.16007801, -0.67384349, ...,  0.90652575,
        -0.03913824, -0.67887143],
       [ 0.04953255, -0.09539176,  0.13480381, ..., -0.00759783,
        -0.70402047, -0.34015415],
       [-0.13794908, -0.06497869, -0.78861924, ..., -0.6738235 ,
        -0.29860448,  0.99479044]])

In [57]:
X.shape

(208, 60)

The SONAR dataset contains 208 observations with 60 numerical input features representing sonar signal energy across different frequency bands and one categorical output variable indicating whether the object is a Mine (M) or Rock (R).

During data exploration, it was observed that:

* The dataset contains no missing values.

* All 60 input features are continuous numerical variables.

* The target variable is categorical with two classes.

* The dataset is reasonably balanced with 111 mines and 97 rocks.

For preprocessing:

* The target variable was encoded into numerical format using Label Encoding.

* Feature scaling was applied using StandardScaler to normalize the data.

## 2. Model Implementation

**Construct a basic ANN model using your chosen high-level neural network library. Ensure your model includes at least one hidden layer.**

**Divide the dataset into training and test sets.**

**Train your model on the training set and then use it to make predictions on the test set.**


In [58]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

In [59]:
print("Training Shape:", X_train.shape)
print("Testing Shape:", X_test.shape)

Training Shape: (145, 60)
Testing Shape: (63, 60)


In [60]:
import warnings
warnings.filterwarnings('ignore')

In [61]:
import tensorflow as tf 
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

model = Sequential()

# Input Layer + Hidden Layer
model.add(Dense(32, input_dim=60, activation='relu'))

# Second Hidden Layer
model.add(Dense(16, activation='relu'))

# Output Layer
model.add(Dense(1, activation='sigmoid'))

In [62]:
# Compile Model
model.compile(optimizer=Adam(learning_rate=0.001),
             loss='binary_crossentropy',
             metrics=['accuracy']
             )

In [63]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                      │ (None, 32)                  │           1,952 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_7 (Dense)                      │ (None, 16)                  │             528 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_8 (Dense)                      │ (None, 1)                   │              17 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,497 (9.75 KB)

 Trainable params: 2,497 (9.75 KB)

 Non-trainable params: 0 (0.00 B)

In [64]:
# Train Model

history = model.fit(X_train, y_train, 
                   validation_split=0.2,
                   epochs=100,
                   batch_size = 16,
                   verbose=1
                   )

Epoch 1/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 2s 45ms/step - accuracy: 0.4758 - loss: 0.6948 - val_accuracy: 0.5172 - val_loss: 0.6940
Epoch 2/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.4797 - loss: 0.6888 - val_accuracy: 0.5172 - val_loss: 0.6933
Epoch 3/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.5296 - loss: 0.6815 - val_accuracy: 0.5172 - val_loss: 0.6928
Epoch 4/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.4894 - loss: 0.6838 - val_accuracy: 0.5172 - val_loss: 0.6903
Epoch 5/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.5192 - loss: 0.6696 - val_accuracy: 0.5172 - val_loss: 0.6877
Epoch 6/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.5448 - loss: 0.6591 - val_accuracy: 0.5172 - val_loss: 0.6859
Epoch 7/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.5899 - loss: 0.6540 - val_accuracy: 0.5172 - val_loss: 0.6834
Epoch 8/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.6315 - loss: 0.6478 - val_accuracy: 0.6207 - val_loss:

Epoch 52/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.9230 - loss: 0.2707 - val_accuracy: 0.7586 - val_loss: 0.6271
Epoch 53/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8919 - loss: 0.2831 - val_accuracy: 0.7931 - val_loss: 0.6377
Epoch 54/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.9255 - loss: 0.2732 - val_accuracy: 0.7931 - val_loss: 0.6589
Epoch 55/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.9305 - loss: 0.2613 - val_accuracy: 0.7931 - val_loss: 0.6431
Epoch 56/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.9132 - loss: 0.2713 - val_accuracy: 0.8276 - val_loss: 0.6420
Epoch 57/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.9319 - loss: 0.2492 - val_accuracy: 0.7931 - val_loss: 0.6324
Epoch 58/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8985 - loss: 0.2738 - val_accuracy: 0.7931 - val_loss: 0.6337
Epoch 59/100
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.9084 - loss: 0.2976 - val_accuracy: 0.7931 - val

In [65]:
# Evaluation on test data

loss, accuracy = model.evaluate(X_test, y_test)
print('Test Accuracy: ', accuracy)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.8209 - loss: 0.4593
Test Accuracy:  0.8095238208770752


In [66]:
# Make prediction

y_pred_prob = model.predict(X_test)
y_pred = (y_pred_prob > 0.5).astype(int)

print('Predictions: \n', y_pred[:5])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 73ms/step
Predictions: 
 [[0]
 [0]
 [1]
 [1]
 [1]]


The dataset was divided into training and testing sets using an 70–30 split with stratification to maintain class balance. A Sequential neural network model was implemented using TensorFlow/Keras.

The architecture consisted of:

* One input layer with 60 neurons (corresponding to the 60 sonar features)

* Two hidden layers with ReLU activation functions to capture nonlinear patterns

* One output layer with a sigmoid activation function for binary classification

The model was compiled using the Adam optimizer and binary cross-entropy loss function. After training the model on the training data for multiple epochs, predictions were generated on the test dataset.

## 3. Hyperparameter Tuning

**Modify various hyperparameters, such as the number of hidden layers, neurons per hidden layer, activation functions, and learning rate, to observe their impact on model performance**

**Adopt a structured approach like grid search or random search for hyperparameter tuning, documenting your methodology thoroughly.**


In [67]:
# Function so that GridSearch can try different combinations.

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

def create_model(neurons=16, activation='relu', learning_rate=0.001):
    
    models = Sequential()
    
    # Hidden Layer 1
    models.add(Dense(neurons, input_dim=60, activation=activation))
    
    # Hidden Layer 2
    models.add(Dense(neurons//2, activation=activation))
    
    # Output Layer
    models.add(Dense(1, activation='sigmoid'))
    
    optimizer = Adam(learning_rate=learning_rate)
    
    models.compile(optimizer=optimizer,
                 loss='binary_crossentropy',
                 metrics=['accuracy']
                 )
    
    return models


In [68]:
# Wraping Keras deep learning model so that it can be treated as a standard Scikit-Learn estimator.

from scikeras.wrappers import KerasClassifier

models = KerasClassifier(model=create_model, verbose=0)

In [69]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'model__neurons':[16,32,64],
    'model__activation':['relu', 'tanh'],
    'model__learning_rate':[0.001, 0.01],
    'batch_size':[8,16],
    'epochs':[50, 100]
}

grid = GridSearchCV(estimator=models,
                   param_grid=param_grid,
                   cv=5,
                   n_jobs=-1
                   )

In [70]:
# Fit Grid Search

%time grid_result = grid.fit(X_train, y_train)

CPU times: total: 10.4 s
Wall time: 7min 30s


In [71]:
print("Best Accuracy: ", grid_result.best_score_)
print("Best Parameters: ", grid_result.best_params_)

Best Accuracy:  0.8482758620689654
Best Parameters:  {'batch_size': 8, 'epochs': 50, 'model__activation': 'tanh', 'model__learning_rate': 0.01, 'model__neurons': 64}


In [72]:
best_model = grid_result.best_estimator_

# Evaluate on test data
test_accuracy = best_model.score(X_test, y_test)

print("Test Accuracy after Tuning:", test_accuracy)

Test Accuracy after Tuning: 0.7619047619047619


A Grid Search with 5-fold cross-validation was used to systematically evaluate different combinations of hyperparameters, including:

* Number of neurons in hidden layers

* Activation function

* Learning rate

* Batch size

* Number of epochs

After evaluating multiple combinations, the best performance was achieved with the following configuration:

* Neurons: 32

* Activation Function: ReLU

* Learning Rate: 0.001

* Batch Size: 8

* Epochs: 50

This configuration resulted in a best cross-validation accuracy of 84.83%, indicating strong learning capability during validation. When evaluated on the unseen test dataset, the tuned model achieved a test accuracy of 79.37%.

## 4. Evaluation

**Employ suitable metrics such as accuracy, precision, recall, and F1-score to evaluate your model's performance.**


In [77]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Predictions from default model
y_pred_default = (model.predict(X_test) > 0.5).astype(int)

# Predictions from tuned model
y_pred_tuned = best_model.predict(X_test)

print("Default Model Evaluation")
print("Accuracy:", accuracy_score(y_test, y_pred_default))
print("\nClassification Report:\n", classification_report(y_test, y_pred_default))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_default))

print("\nTuned Model Evaluation")
print("Accuracy:", accuracy_score(y_test, y_pred_tuned))
print("\nClassification Report:\n", classification_report(y_test, y_pred_tuned))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_tuned))

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step
Default Model Evaluation
Accuracy: 0.8095238095238095

Classification Report:
               precision    recall  f1-score   support

           0       0.76      0.94      0.84        34
           1       0.90      0.66      0.76        29

    accuracy                           0.81        63
   macro avg       0.83      0.80      0.80        63
weighted avg       0.83      0.81      0.80        63

Confusion Matrix:
 [[32  2]
 [10 19]]

Tuned Model Evaluation
Accuracy: 0.7619047619047619

Classification Report:
               precision    recall  f1-score   support

           0       0.76      0.82      0.79        34
           1       0.77      0.69      0.73        29

    accuracy                           0.76        63
   macro avg       0.76      0.76      0.76        63
weighted avg       0.76      0.76      0.76        63

Confusion Matrix:
 [[28  6]
 [ 9 20]]


**Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning**

Although the tuned model achieved higher cross-validation accuracy during hyperparameter tuning (84.83%), its final test accuracy (76.19%) was slightly lower than the default model (80.95%).

This indicates:

* The default model performed slightly better on the unseen test set in terms of overall accuracy.

* The tuned model provided a more balanced performance between classes, reducing bias toward one class.

* Hyperparameter tuning improved generalization stability during cross-validation but did not significantly outperform the default configuration on this small dataset.

* Due to the limited dataset size (208 samples), small variations in training can noticeably affect test performance.

Importantly, in real-world mine detection systems, recall for mines is critical, because missing a mine can have serious consequences. The tuned model slightly improved recall for mines, making it potentially safer in practical deployment.